### PR TITLE
test: added unit and integration tests for Product and Category APIs 

### PR DESCRIPTION
**Pull request by GunjanTewari1404**
- added unit tests for the service layers CategoryService and ProductService by mocking their repository layers.
- added a seed script for seeding sample data for categories and products to the MongoDB test database which is used by integration tests.
- added pytest fixtures for integration test setup with MongoDB, test data seeding and API client initialization.
- added integration tests for product and category APIs covering all the edge cases and tested the end-to-end flow for category-product management.

### CODE REVIEW

File: product/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import pytest
+import mongoengine
```
Comment (Aryan-Satija) [inline product/tests/conftest.py]: Instead of importing entire module, we should prefer to import only specific functions from the module which are being used.
`from mongoengine import connect, disconnect`

File: product/tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#seed test MONGODB with categ. and prod for each test
+@pytest.fixture(scope="function")
+def test_data(mongodb_connection):
+    data=seed_test_database()
+    yield data
+    clear_test_data()
+
+@pytest.fixture
+def api_client():
+    from rest_framework.test import APIClient
```
Comment (Aryan-Satija) [inline product/tests/conftest.py]: why haven't we imported it at the top of the file?

File: product/tests/test_category_unit.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    CATEGORY_ID=ObjectId()
+    PRODUCT_ID=ObjectId()
+    
+    def setUp(self):
+        self.mock_category_repo=Mock()            #create mock repo
+        self.category_service=CategoryService()   #create service with mock repo
+        self.category_service.repository=self.mock_category_repo
+        self.sample_categ_data={'title': 'Test Category','description': 'This is a test category'}     #sample categ. data for testing
+        self.sample_categ=Mock(spec=ProductCategory)   #sample categ.
+        self.sample_categ.id=ObjectId(self.CATEGORY_ID)
```
Comment (Aryan-Satija) [inline product/tests/test_category_unit.py]: CATEGORY_ID is already an ObjectId.
I guess, We can directly assign it like this:
`self.sample_categ.id = self.CATEGORY_ID ` should work as well.
Applicable everywhere.

File: product/tests/test_product_unit.py
Comment (Aryan-Satija) [inline product/tests/test_product_unit.py:1]: We can also add testcases covering other edge cases like

1. Trying to create a product with negative price
2. Trying to create a product with negative quantity
3. Trying to create a product with incomplete fields (no name or no price), etc

File: product/tests/test_category_unit.py (hunk 3 above)
Comment (GunjanTewari1404) [reply product/tests/test_category_unit.py]: Yes, this can be updated for sure.

File: product/tests/conftest.py (hunk 2 above)
Comment (GunjanTewari1404) [reply product/tests/conftest.py]: The import was required only within the scope of this fixture. But moving it up would make the code more readable.

File: product/repository/ProductCategoryRepo.py
Comment (JatinSH) [inline product/repository/ProductCategoryRepo.py:1]: I see everything as commented in this file. Are these some temporary changes?

File: product/migrations/update_brand.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from django.core.management.base import BaseCommand
+from ..models.ProductModel import Product
+
+class Command(BaseCommand):
+    help='Add a default brand to products without a brand'
+
+    def handle(self, *args, **kwargs):
+        prod_without_brand=Product.objects.filter(brand__in=[None, ''])
```
Comment (JatinSH) [inline product/migrations/update_brand.py:8]: Something to think about : what will happen if there were products which didn't even have the brand field?

File: product/scripts/seed_testDB.py
Comment (JatinSH) [inline product/scripts/seed_testDB.py:1]: Great job on the seed script 👏

File: product/services/CategoryService.py [hunk 5]
```diff
@@ -32,7 +33,34 @@ def get_products_in_category(self, category_id):
         return Product.objects.filter(category=category)
     
     def add_prod_to_category(self,category_id,product_id):
-        return self.repository.add_prod_to_category(category_id,product_id)
+       try:
+            category=ProductCategory.objects.get(id=category_id)
+            product=Product.objects.get(id=product_id)
```
Comment (JatinSH) [inline product/services/CategoryService.py]: Instead of directly querying your DB in. the service, you should use the repository layer here. 

```
category = self.product_category_repo. find_by_id(id=category_id)
product = self.product_repo.find_by_id(id= product_id)
```

Your service layer should not access the database directly.

File: product/services/CategoryService.py [hunk 6]
```diff
@@ -32,7 +33,34 @@ def get_products_in_category(self, category_id):
...
+            
+            #check if product is already in this category
+            if product.category == category:
+                return False
+            
+            #update product's category
+            product.category=category
+            product.save()
```
Comment (JatinSH) [inline product/services/CategoryService.py]: This update should also happen through repo layer.
`self.product_repo.update(id= product_id,category=category)`

File: product/services/CategoryService.py [hunk 7]
```diff
@@ -32,7 +33,34 @@ def get_products_in_category(self, category_id):
...
+            return True
+       except Exception:  # Catch all exceptions
+            return False
+        # return self.repository.add_prod_to_category(category_id,product_id)
 
     def remove_prod_from_category(self,product_id):
-        return self.repository.remove_prod_from_category(product_id)
\ No newline at end of file
+        try:
+            product=Product.objects.get(id=product_id)
```
Comment (JatinSH) [inline product/services/CategoryService.py]: Same recommendation as above here too

File: product/tests/conftest.py (hunk 1 above)
Comment (JatinSH) [reply product/tests/conftest.py]: +1 to that

File: product/tests/end_to_end.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
+from django.urls import reverse
+from product.models.ProductModel import Product
+from bson import ObjectId
+
+class TestProductCategoryIntegration:
```
Comment (JatinSH) [inline product/tests/end_to_end.py:5]: @GunjanTewari1404 Is this working locally? I'm curious to know who is supplying the `api_client, test_data` to the tests in this file

File: product/tests/end_to_end.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        #step 4 - move prod to new categ.
+        management_url = reverse('category-product-management')
+        move_payload = {
+            "product_id": product_id,
+            "category_id": another_category_id
+        }
+
+        move_res = api_client.post(management_url, move_payload, format='json')
+        assert move_res.status_code == 200
+        assert any(key in move_res.json() for key in ("success", "message"))
```
Comment (JatinSH) [inline product/tests/end_to_end.py]: Just a suggestion. If you APIs return proper error/success messages, you can assert them here

File: product/tests/end_to_end.py
Comment (JatinSH) [inline product/tests/end_to_end.py:1]: Awesome job on the end to end integration test

File: product/tests/conftest.py (hunk 1 above)
Comment (GunjanTewari1404) [reply product/tests/conftest.py]: Done! Have optimised the imports accordingly.

File: product/repository/ProductCategoryRepo.py
Comment (GunjanTewari1404) [reply product/repository/ProductCategoryRepo.py:1]: Yes I had commented those two functions temporarily while working on related changes. Fixed the logic now.

File: product/migrations/update_brand.py (hunk 4 above)
Comment (GunjanTewari1404) [reply product/migrations/update_brand.py:8]: Thank you for bringing this up! I have added the logic to handle the cases where the 'brand' field might be missing using 'hasattr()'.

File: product/scripts/seed_testDB.py
Comment (GunjanTewari1404) [reply product/scripts/seed_testDB.py:1]: Thank you!

File: product/services/CategoryService.py (hunk 5 above)
Comment (GunjanTewari1404) [reply product/services/CategoryService.py]: I have updated the service to keep only the logic. The DB calls are handled through the repository layer.

File: product/services/CategoryService.py (hunk 6 above)
Comment (GunjanTewari1404) [reply product/services/CategoryService.py]: Updated! Moved the update logic into the repository layer using self.product_repo.update() as suggested.
```
if not category or not product:
            return False
        if product.category == category:
            return False
        updated = self.product_repo.update(product_id, {"category": category})
        return updated is not None
```

File: product/services/CategoryService.py (hunk 7 above)
Comment (GunjanTewari1404) [reply product/services/CategoryService.py]: Updated this too. Now using the repo’s get_by_id instead of querying directly. 
```
def remove_prod_from_category(self,product_id):
        product = self.product_repo.get_by_id(product_id)
        if not product or not product.category:
            return False
```

File: product/tests/end_to_end.py (hunk 8 above)
Comment (GunjanTewari1404) [reply product/tests/end_to_end.py:5]: Yes it is working succesfully in the local. The api_client, test_data to the tests in this file are provided throgh pytest fixtures in conftest.py file and pytest automatically injects them when the tests are run.

File: product/tests/end_to_end.py (hunk 9 above)
Comment (GunjanTewari1404) [reply product/tests/end_to_end.py]: Thank you for the suggestion! I’ve included assertions for the success/error messages in the test cases.

```
 move_res = api_client.post(management_url, move_payload, format='json')
 assert move_res.status_code == 200
 data = move_res.json()
 assert "message" in data
 assert data["message"] == "Product added to category successfully"
```

File: product/tests/end_to_end.py
Comment (GunjanTewari1404) [reply product/tests/end_to_end.py:1]: Thank you for the feedback!

File: product/tests/test_product_unit.py
Comment (GunjanTewari1404) [reply product/tests/test_product_unit.py:1]: Thank you for the feedback! I have already added validations for scenarios like 1,2. I have covered the remaining edge cases in integration testing.

File: product/tests/end_to_end.py (hunk 8 above)
Comment (JatinSH) [reply product/tests/end_to_end.py:5]: Okay great. Glad to know

**Review by JatinSH — APPROVED** [review]
Looks good to me! Great job.

**Review by Aryan-Satija — APPROVED** [review]
Looks good to me!
